File: 2/dayTwo.R
library(tidyverse)
library(rlist)
library(doParallel)

### Part One
inputDay2 <- read_csv("~/Downloads/inputDay2.txt", col_types = "c", col_names = "direction")
data <- inputDay2$direction %>%
  map(str_split, pattern = " ") %>%
  list.flatten() %>%
  tibble(direction = map_chr(., 1),
         amount = map_chr(., 2) %>%
           as.integer()) %>%
  select(direction, amount)

group_by(data, direction) %>%
  summarize(value = sum(amount))

## This is horrible and I shouldn't be doing manual calculations like this but oh well. Quick and dirty.
(1910 - 977) * 2165

### Part Two
inputDayTwo <- read_lines("~/Downloads/inputDay2.txt")
aim = 0
position = list(horiontal = 0, depth = 0)
foreach(string = inputDayTwo, .combine = bind_rows) %do% {
  direction <- str_split(string, " ")[[1]][1]
  amount <- str_split(string, " ")[[1]][2] %>% as.numeric()
    if(direction == "down") { aim <<- aim + amount }
    if(direction == "up") { aim <<- aim - amount }
    if(direction == "forward") {
      position$horiontal <<- position$horiontal + amount
      position$depth <<- position$depth + 2*aim
    }
  tibble(direction, amount)
}

File: 2/dayTwo.py
def pos(string):
  '''
  String is the Elf submarine command.
  Aim is the present aim, passed by reference.
  Poisition is the list: [horizontal, depth]; passed by reference also.
  '''
  
  str = string.split()
  
  global aim
  global position
  
  if str[0][0] != 'f':
    if str[0][0] == 'd':
      print("Command is not: `forward`.")
      print("Aim before command:", aim)
      aim += int(str[1])  # Increment aim, increasing depth.
      print("Aim after command:", aim)
    else:
      aim -= int(str[1])  # Decrement aim, decreasing depth.
  else:
    print("Command is: `forward`.")
    print("Position before command:", position)
    position = [position[0] + int(str[1]),
                position[1] + int(str[1]) * aim]
    print("Position after command:", position)
  
  # Void: aim, position.

with open("/Users/bryce/Downloads/inputDay2.txt") as d:
  data = d.readlines()

aim = 0
position = [0,0]

for command in data:
  pos(command)

print(position[0] * position[1])
